File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2016/Y16D19.kt
package org.gristle.pdxpuzzles.advent.y2016

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.math.pow
import kotlin.math.log10

class Y16D19(input: String) : Day {
    private val elves = input.toInt()

    override fun part1(): Int {
        val exponent = (log10(elves.toDouble()) / log10(2.0)).toInt()
        return (elves - (2.pow(exponent).toInt())) * 2 + 1
    }

    override fun part2(): Int {
        val exponent = (log10(elves.toDouble()) / log10(3.0)).toInt()
        val lastUp = 3.pow(exponent).toInt()
        val diff = elves - 3.pow(exponent).toInt()
        val ones = minOf(diff, lastUp)
        val twos = maxOf(diff - ones, 0)
        return ones + twos * 2
    }
}

// Class creation: 5ms
// Part 1: 1816277 (0ms)
// Part 2: 1410967 (0ms)
fun main() = Day.runDay(Y16D19::class) 

File: rust/advent/src/bin/2016/y16d19.rs
use std::cmp::{max, min};

use advent::utilities::get_input::get_input;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = usize;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(16, 19).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(input), stopwatch.lap().report());
    println!("2. {} ({})", part2(input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input.parse().unwrap()
}

fn part1(elves: Input) -> Output {
    let exponent = (f64::log10(elves as f64) / f64::log10(2.0)) as u32;
    (elves - 2usize.pow(exponent)) * 2 + 1
}

fn part2(elves: Input) -> Output {
    let exponent = (f64::log10(elves as f64) / f64::log10(3.0)) as u32;
    let last_up = 3usize.pow(exponent);
    let diff = elves - last_up;
    let ones = min(diff, last_up);
    let twos = max(diff - ones, 0);
    ones + twos * 2
}

#[test]
fn default() {
    let input = get_input(16, 19).unwrap();
    let input = parse_input(&input);
    assert_eq!(1816277, part1(input));
    assert_eq!(1410967, part2(input));
}

// Input parsed (18μs)
// 1. 1816277 (12μs)
// 2. 1410967 (2μs)
// Total: 35μs
